src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Completion/CompletionListCache.cs

```diff
@@ -2,28 +2,30 @@
 // Licensed under the MIT license. See License.txt in the project root for license information.
 
 using System;
-using System.Collections.Generic;
-using System.Diagnostics.CodeAnalysis;
 using Microsoft.VisualStudio.LanguageServer.Protocol;
 
 namespace Microsoft.AspNetCore.Razor.LanguageServer.Completion;
 
 internal sealed class CompletionListCache
 {
+    private record struct Slot(
+        int Id,
+        VSInternalCompletionList CompletionList,
+        object? Context,
+        bool Used = true);
+
     // Internal for testing
-    internal static readonly int MaxCacheSize = 10;
+    internal const int MaxCacheSize = 10;
 
-    private readonly object _accessLock;
-    private readonly List<CacheEntry> _resultIdToCacheEntry;
-    private int _nextResultId;
+    private readonly object _accessLock = new();
 
-    public CompletionListCache()
-    {
-        _accessLock = new object();
-        _resultIdToCacheEntry = new List<CacheEntry>();
-    }
+    // This is used as a circular buffer.
+    private readonly Slot[] _items = new Slot[MaxCacheSize];
 
-    public int Set(VSInternalCompletionList completionList, object? context)
+    private int _nextIndex;
```

Comment:
It might be a bit cleaner to move circular buffer into its own class and have read/write logic there.

Reply:
Can you describe how you imagine that being cleaner? In my mind, the entire implementation of this class is the implementation of a circular buffer, so pushing that into another class seems redundant. The other class would be just as readable (or not). I could create a new generic collection class to represent a circular buffer, but that doesn't seem all that useful at the moment (i.e., YAGNI). After all, it's just an array and an int.

Reply:
I think it was naming that was rubbing me the wrong way, but looking at the full implementation after checking out it makes some sense. Somewhere it is weird to me that we have basically a wrapper around a circular buffer to make it a singleton, and I wish there was a more elegant way to make that without making a fully separate class. 

After toying with it a few minutes I agree, not as clean as my hunch was. If we were going to do anything I think it would be to make this just a generic circular buffer class and have completion depend on it, but I didn't realize that completion had this a singleton dependency injection.

Reply:
Totally agree that it's a bit weird, but there are a few places where that singleton `CompletionListCache` is shared via DI.

Reply:
I initially had the same thought, about making a circular buffer class, but came to the same conclusion as Dustin: this is essentially that, just with a questionable name, and if we ever need one again, we can do the work then.

Reply:
I wonder why the cache works like that. Is it just because of all the different provides? Shouldn't each provider have its own cache?

Reply:
I'm not 100% sure, but I intend to take a look! :smile: :+1: